### ISSUE TITLE
Reproduce code issues

### ISSUE THREAD

**Issue by Arbor334** [PKU-YuanGroup/Chat-UniVi#60]
Hello, I used phi to reproduce your model, but encountered a list out of range issue while loading the model. How can I solve this problem? Here is the data after fine-tuning
![issue1](https://github.com/user-attachments/assets/f409851f-5ca2-4b9d-912e-4c0d1c5dff15)

![issue](https://github.com/user-attachments/assets/4edc22df-f6c5-4453-b64c-bd90ae2dd707)

**Comment by jpthu17** [comment]
Sorry, for now, we cannot use Transformers.AutoModelForCausaLLM to load our model. You can load the model like this:
https://github.com/PKU-YuanGroup/Chat-UniVi?tab=readme-ov-file#inference-for-video-understanding

**Comment by Arbor334** [comment]
Thank you for your reply. I got out of bed. Please wait a moment, don't go away. I have some questions that need to be explained to you

**Comment by Arbor334** [comment]
------------------------- ------------
accelerate                0.21.0
addict                    2.4.0
aiofiles                  24.1.0
aiohappyeyeballs          2.4.0
aiohttp                   3.10.5
aiosignal                 1.3.1
altair                    5.4.1
anyio                     4.4.0
async-timeout             4.0.3
attrs                     24.2.0
bitsandbytes              0.41.0
certifi                   2024.8.30
charset-normalizer        3.3.2
click                     8.1.7
cmake                     3.30.2
contourpy                 1.3.0
cycler                    0.12.1
decord                    0.6.0
deepspeed                 0.9.5
distro                    1.9.0
einops                    0.6.1
exceptiongroup            1.2.2
fastapi                   0.112.2
ffmpy                     0.4.0
filelock                  3.15.4
flash-attn                2.5.2
fonttools                 4.53.1
frozenlist                1.4.1
fsspec                    2024.6.1
gradio                    3.23.0
h11                       0.14.0
hjson                     3.1.0
httpcore                  1.0.5
httpx                     0.27.2
huggingface-hub           0.24.6
idna                      3.8
importlib_metadata        8.4.0
Jinja2                    3.1.4
jiter                     0.5.0
jsonschema                4.23.0
jsonschema-specifications 2023.12.1
kiwisolver                1.4.7
linkify-it-py             2.0.3
lit                       18.1.8
markdown-it-py            2.2.0
markdown2                 2.4.8
MarkupSafe                2.1.5
matplotlib                3.9.2
mdit-py-plugins           0.3.3
mdurl                     0.1.2
mmengine                  0.10.4
mpmath                    1.3.0
multidict                 6.0.5
narwhals                  1.6.2
networkx                  3.3
ninja                     1.11.1.1
numpy                     1.26.4
nvidia-cublas-cu11        11.11.3.6
nvidia-cuda-cupti-cu11    11.8.87
nvidia-cuda-nvrtc-cu11    11.8.89
nvidia-cuda-runtime-cu11  11.8.89
nvidia-cudnn-cu11         9.1.0.70
nvidia-cufft-cu11         10.9.0.58
nvidia-curand-cu11        10.3.0.86
nvidia-cusolver-cu11      11.4.1.48
nvidia-cusparse-cu11      11.7.5.86
nvidia-nccl-cu11          2.20.5
nvidia-nvtx-cu11          11.8.86
openai                    1.44.0
opencv-python             4.10.0.84
orjson                    3.10.7
packaging                 24.1
pandas                    2.2.2
peft                      0.12.0
Pillow                    9.5.0
pip                       22.3.1
platformdirs              4.2.2
protobuf                  4.23.2
psutil                    6.0.0
py-cpuinfo                9.0.0
pydantic                  1.10.7
pydub                     0.25.1
Pygments                  2.18.0
pyparsing                 3.1.4
python-dateutil           2.9.0.post0
python-multipart          0.0.9
pytz                      2024.1
PyYAML                    6.0.2
referencing               0.35.1
regex                     2024.7.24
requests                  2.30.0
rich                      13.8.0
rpds-py                   0.20.0
safetensors               0.4.4
scipy                     1.14.1
semantic-version          2.10.0
sentencepiece             0.1.99
setuptools                65.5.1
six                       1.16.0
sniffio                   1.3.1
starlette                 0.38.4
sympy                     1.13.2
termcolor                 2.4.0
timm                      1.0.9
tokenizers                0.19.1
tomli                     2.0.1
torch                     2.1.2+cu118
torchvision               0.16.2+cu118
tqdm                      4.65.0
transformers              4.41.1
triton                    2.1.0
typing_extensions         4.12.2
tzdata                    2024.1
uc-micro-py               1.0.3
urllib3                   2.2.2
uvicorn                   0.30.6
websockets                13.0.1
wheel                     0.38.4
yapf                      0.40.2
yarl                      1.9.8
zipp                      3.20.1

**Comment by Arbor334** [comment]
I conducted the test using your data on HF, and it looked like this, and then the answer appeared

**Comment by Arbor334** [comment]
import torch
import os
from ChatUniVi.constants import *
from ChatUniVi.conversation import conv_templates, SeparatorStyle
from ChatUniVi.model.builder import load_pretrained_model
from ChatUniVi.utils import disable_torch_init
from ChatUniVi.mm_utils import tokenizer_image_token, get_model_name_from_path, KeywordsStoppingCriteria
from PIL import Image
from decord import VideoReader, cpu
import numpy as np


def _get_rawvideo_dec(video_path, image_processor, max_frames=MAX_IMAGE_LENGTH, image_resolution=224, video_framerate=1, s=None, e=None):
    # speed up video decode via decord.

    if s is None:
        start_time, end_time = None, None
    else:
        start_time = int(s)
        end_time = int(e)
        start_time = start_time if start_time >= 0. else 0.
        end_time = end_time if end_time >= 0. else 0.
        if start_time > end_time:
            start_time, end_time = end_time, start_time
        elif start_time == end_time:
            end_time = start_time + 1

    if os.path.exists(video_path):
        vreader = VideoReader(video_path, ctx=cpu(0))
    else:
        print(video_path)
        raise FileNotFoundError

    fps = vreader.get_avg_fps()
    f_start = 0 if start_time is None else int(start_time * fps)
    f_end = int(min(1000000000 if end_time is None else end_time * fps, len(vreader) - 1))
    num_frames = f_end - f_start + 1
    if num_frames > 0:
        # T x 3 x H x W
        sample_fps = int(video_framerate)
        t_stride = int(round(float(fps) / sample_fps))

        all_pos = list(range(f_start, f_end + 1, t_stride))
        if len(all_pos) > max_frames:
            sample_pos = [all_pos[_] for _ in np.linspace(0, len(all_pos) - 1, num=max_frames, dtype=int)]
        else:
            sample_pos = all_pos

        patch_images = [Image.fromarray(f) for f in vreader.get_batch(sample_pos).asnumpy()]

        patch_images = torch.stack([image_processor.preprocess(img, return_tensors='pt')['pixel_values'][0] for img in patch_images])
        slice_len = patch_images.shape[0]

        return patch_images, slice_len
    else:
        print("video path: {} error.".format(video_path))


if __name__ == '__main__':
    # Model Parameter
    model_path = "Chat-Univi/Chat-Univi-7B"  # or "Chat-UniVi/Chat-UniVi-13B"
    video_path = "/workspace/dataset/private/ZiJian/chatunivi_data/instruction_tuning/Activity_Videos/v_zzxYEZkahBU.mp4"

    # The number of visual tokens varies with the length of the video. "max_frames" is the maximum number of frames.
    # When the video is long, we will uniformly downsample the video to meet the frames when equal to the "max_frames".
    max_frames = 100

    # The number of frames retained per second in the video.
    video_framerate = 1

    # Input Text
    qs = "Describe the video."

    # Sampling Parameter
    conv_mode = "simple"
    temperature = 0.2
    top_p = None
    num_beams = 1

    disable_torch_init()
    model_path = os.path.expanduser(model_path)
    model_name = "ChatUniVi"
    tokenizer, model, image_processor, context_len = load_pretrained_model(model_path, None, model_name)

    mm_use_im_start_end = getattr(model.config, "mm_use_im_start_end", False)
    mm_use_im_patch_token = getattr(model.config, "mm_use_im_patch_token", True)
    if mm_use_im_patch_token:
        tokenizer.add_tokens([DEFAULT_IMAGE_PATCH_TOKEN], special_tokens=True)
    if mm_use_im_start_end:
        tokenizer.add_tokens([DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN], special_tokens=True)
    model.resize_token_embeddings(len(tokenizer))

    vision_tower = model.get_vision_tower()
    if not vision_tower.is_loaded:
        vision_tower.load_model()
    image_processor = vision_tower.image_processor

    if model.config.config["use_cluster"]:
        for n, m in model.named_modules():
            m = m.to(dtype=torch.bfloat16)

    # Check if the video exists
    if video_path is not None:
        video_frames, slice_len = _get_rawvideo_dec(video_path, image_processor, max_frames=max_frames, video_framerate=video_framerate)

        cur_prompt = qs
        if model.config.mm_use_im_start_end:
            qs = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN * slice_len + DEFAULT_IM_END_TOKEN + '\n' + qs
        else:
            qs = DEFAULT_IMAGE_TOKEN * slice_len + '\n' + qs

        conv = conv_templates[conv_mode].copy()
        conv.append_message(conv.roles[0], qs)
        conv.append_message(conv.roles[1], None)
        prompt = conv.get_prompt()

        input_ids = tokenizer_image_token(prompt, tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt').unsqueeze(
            0).cuda()

        stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
        keywords = [stop_str]
        stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)

        with torch.inference_mode():
            output_ids = model.generate(
                input_ids,
                images=video_frames.half().cuda(),
                do_sample=True,
                temperature=temperature,
                top_p=top_p,
                num_beams=num_beams,
                output_scores=True,
                return_dict_in_generate=True,
                max_new_tokens=1024,
                use_cache=True,
                stopping_criteria=[stopping_criteria])

        output_ids = output_ids.sequences
        input_token_len = input_ids.shape[1]
        n_diff_input_output = (input_ids != output_ids[:, :input_token_len]).sum().item()
        if n_diff_input_output > 0:
            print(f'[Warning] {n_diff_input_output} output_ids are not the same as the input_ids')
        outputs = tokenizer.batch_decode(output_ids[:, input_token_len:], skip_special_tokens=True)[0]
        outputs = outputs.strip()
        if outputs.endswith(stop_str):
            outputs = outputs[:-len(stop_str)]
        outputs = outputs.strip()
        print(outputs)

**Comment by Arbor334** [comment]
![Uploading Snipaste_2024-09-24_14-11-01.png…]()

**Comment by Arbor334** [comment]
But I encountered this problem when I trained myself

**Comment by Arbor334** [comment]
![Uploading Snipaste_2024-09-24_14-12-20.png…]()

**Comment by Arbor334** [comment]
![Snipaste_2024-09-24_14-14-13](https://github.com/user-attachments/assets/33eb419f-584a-4390-807a-9d75618444d3)

**Comment by Arbor334** [comment]
Do I need to do anything after fine-tuning

**Comment by Arbor334** [comment]
I really need your help

**Comment by jpthu17** [comment]
The images seem to fail to upload?
![image](https://github.com/user-attachments/assets/fc0c7c9a-b7e9-4b02-91c6-331445bab61e)

**Comment by Arbor334** [comment]
Updated, please check

**Comment by Arbor334** [comment]
![1](https://github.com/user-attachments/assets/06e221c5-70c1-4e9e-8406-505d3e8f78f5)
![2](https://github.com/user-attachments/assets/9b392bc5-586b-4259-a223-2f38d0faf3e0)
![3](https://github.com/user-attachments/assets/c1a30c2d-8353-456a-aae9-e4f84e57d656)
During training, the content of the file was changed to ChatUniViPhiForCauselLM

**Comment by Arbor334** [comment]
![4](https://github.com/user-attachments/assets/86a13429-d0ed-452c-90b2-4cf77fe0a848)

**Comment by Arbor334** [comment]
Hello, are you still there

**Comment by jpthu17** [comment]
It seems you didn't initialize config.
```
kwargs = {"device_map": device_map}
model = ChatUniViPhiForCausalLM.from_pretrained(model_base, low_cpu_mem_usage=True, config=cfg_pretrained, **kwargs)
```
Read more about "device_map" here: [Handling big models for inference 58](https://huggingface.co/docs/accelerate/concept_guides/big_model_inference#the-devicemap).

**Comment by Arbor334** [comment]
Can you send me how to conduct experiments using phi. Thank you very much.

**Comment by jpthu17** [comment]
I have not experimented with Phi models. All my experiments were based on LLaMA's architecture. 

However, many people have inquired about the code for Phi. I wrote a portion of the code, and it appeared to be trainable on Phi. However, I did not try the full training and evaluation process.

**Comment by Arbor334** [comment]
Thank you for your response. As a newcomer to learning big models, I have not yet encountered some details. Thanks again
